satellite/satellitedb/offers.go

```diff
@@ -0,0 +1,157 @@
+	statement = `
+		WITH o AS (
+			SELECT ` + columns + ` FROM offers WHERE status=? AND expires_at>?
+		)
+		SELECT ` + columns + ` FROM o
+		UNION ALL
+		SELECT ` + columns + ` FROM offers
+		WHERE status=?
+		AND NOT EXISTS (
+			SELECT id FROM o
+		) order by created_at desc;`
+
+	rows := offers.db.DB.QueryRowContext(ctx, offers.db.Rebind(statement), offerStatus, time.Now(), marketing.Default)
+
+	o := marketing.Offer{}
+	err := rows.Scan(&o.ID, &o.Name, &o.Description, &o.AwardCreditInCents, &o.InviteeCreditInCents, &o.AwardCreditDurationDays, &o.InviteeCreditDurationDays, &o.RedeemableCap, &o.NumRedeemed, &o.ExpiresAt, &o.CreatedAt, &o.Status)
+	if err == sql.ErrNoRows {
+		return nil, marketing.OffersErr.New("no current offer")
+	}
+	if err != nil {
+		return nil, marketing.OffersErr.Wrap(err)
+	}
+
+	return &o, nil
+}
+
+// Create inserts a new offer into the db
+func (offers *offers) Create(ctx context.Context, o *marketing.NewOffer) (*marketing.Offer, error) {
+	tx, err := offers.db.Open(ctx)
+	if err != nil {
+		return nil, marketing.OffersErr.Wrap(err)
+	}
+
+	statement := offers.db.Rebind(`
+		UPDATE offers SET status=?, expires_at=?
+		WHERE status=? AND expires_at>?;
+	`)
+	currentTime := time.Now()
+	_, err = tx.Tx.ExecContext(ctx, statement, marketing.Done, currentTime, o.Status, currentTime)
+	if err != nil {
+		return nil, marketing.OffersErr.Wrap(errs.Combine(err, tx.Rollback()))
+	}
+
+	offerDbx, err := tx.Create_Offer(ctx,
+		dbx.Offer_Name(o.Name),
+		dbx.Offer_Description(o.Description),
+		dbx.Offer_AwardCreditInCents(o.AwardCreditInCents),
+		dbx.Offer_InviteeCreditInCents(o.InviteeCreditInCents),
+		dbx.Offer_AwardCreditDurationDays(o.AwardCreditDurationDays),
+		dbx.Offer_InviteeCreditDurationDays(o.InviteeCreditDurationDays),
+		dbx.Offer_RedeemableCap(o.RedeemableCap),
+		dbx.Offer_ExpiresAt(o.ExpiresAt),
+		dbx.Offer_Status(int(o.Status)),
+	)
+	if err != nil {
+		return nil, marketing.OffersErr.Wrap(errs.Combine(err, tx.Rollback()))
+	}
+
+	newOffer, err := convertDBOffer(offerDbx)
+	if err != nil {
+		return nil, marketing.OffersErr.Wrap(errs.Combine(err, tx.Rollback()))
+	}
+
+	return newOffer, marketing.OffersErr.Wrap(tx.Commit())
+}
+
+// Update modifies an offer entry's status and amount of offers redeemed based on offer id
+func (offers *offers) Update(ctx context.Context, o *marketing.UpdateOffer) error {
+	updateFields := dbx.Offer_Update_Fields{
+		Status:      dbx.Offer_Status(int(o.Status)),
+		NumRedeemed: dbx.Offer_NumRedeemed(o.NumRedeemed),
+		ExpiresAt:   dbx.Offer_ExpiresAt(o.ExpiresAt),
+	}
+
+	offerID := dbx.Offer_Id(o.ID)
+
+	_, err := offers.db.Update_Offer_By_Id(ctx, offerID, updateFields)
+	if err != nil {
+		return marketing.OffersErr.Wrap(err)
+	}
+
+	return nil
+}
+
+func offersFromDbx(offersDbx []*dbx.Offer) ([]marketing.Offer, error) {
+	var offers []marketing.Offer
+	var errors []error
```

Comment:
this is fine, but an optional thing that's also used in the repo is `var errList errs.Group`, where you can add errors like `errList.Add(err)`, then return all with `errList.Err()`. I'm not sure of its differences but i've just seen errs.Group more often.